README.md
# cordova-plugin-mediabeef-background-geolocation


## Description

Cross-platform **background location tracking** plugin for Cordova / PhoneGap with battery-saving "circular region monitoring" and "stop detection".

This plugin can be used for **background location tracking** when app is running in foreground or background. It is more battery and data efficient than html5 geolocation or cordova-geolocation plugin. It can be used side by side with other geolocation providers (eg. html5 navigator.geolocation).
Based on cordova plugin https://github.com/mediabeef/cordova-plugin-background-geolocation

On Android you can choose from two location location providers:
* **ANDROID_DISTANCE_FILTER_PROVIDER**
* **ANDROID_ACTIVITY_PROVIDER**

See [Which provider should I use?](https://github.com/mediabeef/cordova-plugin-background-geolocation/blob/master/PROVIDERS.md) for more information about providers.

## Sample Application

Checkout sample app repository [cordova-plugin-background-geolocation-sample](https://github.com/mediabeef/cordova-plugin-background-geolocation-sample).

## Semantic Versioning
This plugin is following semantic versioning as defined http://semver.org

## Migrations

See [MIGRATIONS.md](/MIGRATIONS.md)

## Installing the plugin

```
cordova plugin add cordova-plugin-mediabeef-background-geolocation
```

**Default iOS location permission prompt can be changed in two ways:**
1. In your config.xml:
    ```
    <plugin name="cordova-plugin-mediabeef-background-geolocation">
        <variable name="LOCATION_ALWAYS_USAGE_DESCRIPTION" value="This app requires background tracking enabled" />
        <variable name="LOCATION_WHEN_IN_USE_USAGE_DESCRIPTION" value="This app requires background tracking enabled" />
        <variable name="LOCATION_ALWAYS_AND_WHEN_IN_USE_USAGE_DESCRIPTION" value="Always use is required for background location tracking" />
    </plugin>
    ```

2. Or, you can pass the following variables on plugin install:
    * `LOCATION_ALWAYS_USAGE_DESCRIPTION`                   for `NSLocationAlwaysUsageDescription`
    * `LOCATION_WHEN_IN_USE_USAGE_DESCRIPTION`              for `NSLocationWhenInUseUsageDescription`
    * `LOCATION_ALWAYS_AND_WHEN_IN_USE_USAGE_DESCRIPTION`   for `NSLocationAlwaysAndWhenInUseUsageDescription`

    Example:
    ```
    cordova plugin add cordova-plugin-mediabeef-background-geolocation --variable LOCATION_ALWAYS_USAGE_DESCRIPTION="Your always usage description" --variable LOCATION_WHEN_IN_USE_USAGE_DESCRIPTION="Your when in use usage description" --variable LOCATION_ALWAYS_AND_WHEN_IN_USE_USAGE_DESCRIPTION="Your always and when in use usage description"
    ```

## Registering plugin for Adobe® PhoneGap™ Build

This plugin should work with Adobe® PhoneGap™ Build without any modification.
To register plugin add following line into your config.xml:

```
<plugin name="cordova-plugin-mediabeef-background-geolocation"/>
```

NOTE: If you're using *hydration*, you have to download and reinstall your app with every new version of the plugin, as plugins are not updated.

## Compilation

### Compatibility

| Plugin version   | Cordova CLI       | Cordova Platform Android | Cordova Platform iOS |
|------------------|-------------------|--------------------------|----------------------|
| <2.3.0           | 6.4.0             | 6.3.0                    |                      |
| >=2.3.0          | 7.1.0             | 6.3.0                    |                      |

**Please note** that as of Cordova Android 6.0.0 icons are by default in mipmap/ directory not drawable/ directory, so this plugin will have a build issue on < 6.0.0 Cordova builds, you will need to update Authenticator.xml to drawable directory from mipmap directory to work on older versions.

### Android
You will need to ensure that you have installed the following items through the Android SDK Manager:

| Name                       | Version |
|----------------------------|---------|
| Android SDK Tools          | 24.4.1  |
| Android SDK Platform-tools | 23.1    |
| Android SDK Build-tools    | 23.0.1  |
| Android Support Repository | 25      |
| Android Support Library    | 23.1.1  |
| Google Play Services       | 29      |
| Google Repository          | 24      |


## Quick Example

```javascript
document.addEventListener('deviceready', onDeviceReady, false);

function onDeviceReady () {

    /**
    * This callback will be executed every time a geolocation is recorded in the background.
    */
    var callbackFn = function(location) {
        console.log('[js] BackgroundGeolocation callback:  ' + location.latitude + ',' + location.longitude);

        // Do your HTTP request here to POST location to your server.
        // jQuery.post(url, JSON.stringify(location));

        /*
        IMPORTANT:  You must execute the finish method here to inform the native plugin that you're finished,
        and the background-task may be completed.  You must do this regardless if your HTTP request is successful or not.
        IF YOU DON'T, ios will CRASH YOUR APP for spending too much time in the background.
        */
        backgroundGeolocation.finish();
    };

    var failureFn = function(error) {
        console.log('BackgroundGeolocation error');
    };

    // BackgroundGeolocation is highly configurable. See platform specific configuration options
    backgroundGeolocation.configure(callbackFn, failureFn, {
        desiredAccuracy: 10,
        stationaryRadius: 20,
        distanceFilter: 30,
        interval: 60000
    });

    // Turn ON the background-geolocation system.  The user will be tracked whenever they suspend the app.
    backgroundGeolocation.start();

    // If you wish to turn OFF background-tracking, call the #stop method.
    // backgroundGeolocation.stop();
}
```

## API

### configure(success, fail, options)

| Parameter | Type          | Platform | Description                                                                     |
|-----------|---------------|----------|---------------------------------------------------------------------------------|
| `success` | `Function`    | all      | Callback to be executed every time a geolocation is recorded in the background. |
| `fail`    | `Function`    | all      | Callback to be executed every time a geolocation error occurs.                  |
| `options` | `JSON Object` | all      | Configure options                                                               |

Configure options:

| Parameter                 | Type              | Platform     | Description                                                                                                                                                                                                                                                                                                                                        |
|---------------------------|-------------------|--------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| `desiredAccuracy`         | `Number`          | all          | Desired accuracy in meters. Possible values [0, 10, 100, 1000]. The lower the number, the more power devoted to GeoLocation resulting in higher accuracy readings. 1000 results in lowest power drain and least accurate readings. @see Apple docs                                                                                                 |
| `stationaryRadius`        | `Number`          | all          | Stationary radius in meters. When stopped, the minimum distance the device must move beyond the stationary location for aggressive background-tracking to engage.                                                                                                                                                                                  |
| `debug`                   | `Boolean`         | all          | When enabled, the plugin will emit sounds for life-cycle events of background-geolocation! See debugging sounds table.                                                                                                                                                                                                                             |
| `distanceFilter`          | `Number`          | all          | The minimum distance (measured in meters) a device must move horizontally before an update event is generated. **@see** [Apple docs](https://developer.apple.com/library/ios/documentation/CoreLocation/Reference/CLLocationManager_Class/CLLocationManager/CLLocationManager.html#//apple_ref/occ/instp/CLLocationManager/distanceFilter).        |
| `stopOnTerminate`         | `Boolean`         | all          | Enable this in order to force a stop() when the application terminated (e.g. on iOS, double-tap home button, swipe away the app). (default true)                                                                                                                                                                                                   |
| `startOnBoot`             | `Boolean`         | Android      | Start background service on device boot. (default false)                                                                                                                                                                                                                                                                                           |
| `startForeground`         | `Boolean`         | Android      | If false location service will not be started in foreground and no notification will be shown. (default true)                                                                                                                                                                                                                                      |
| `interval`                | `Number`          | Android      | The minimum time interval between location updates in milliseconds. **@see** [Android docs](http://developer.android.com/reference/android/location/LocationManager.html#requestLocationUpdates(long,%20float,%20android.location.Criteria,%20android.app.PendingIntent) for more information.                                                     |
| `notificationTitle`       | `String` optional | Android      | Custom notification title in the drawer.                                                                                                                                                                                                                                                                                                           |
| `notificationText`        | `String` optional | Android      | Custom notification text in the drawer.                                                                                                                                                                                                                                                                                                            |
| `notificationIconColor`   | `String` optional | Android      | The accent color to use for notification. Eg. **#4CAF50**.                                                                                                                                                                                                                                                                                         |
| `notificationIconLarge`   | `String` optional | Android      | The filename of a custom notification icon. See android quirks.                                                                                                                                                                                                                                                                                    |
| `notificationIconSmall`   | `String` optional | Android      | The filename of a custom notification icon. See android quirks.                                                                                                                                                                                                                                                                                    |
| `locationProvider`        | `Number`          | Android      | Set location provider **@see** [PROVIDERS.md](https://github.com/mediabeef/cordova-plugin-background-geolocation/blob/master/PROVIDERS.md)                                                                                                                                                                                                          |
| `activityType`            | `String`          | iOS          | [AutomotiveNavigation, OtherNavigation, Fitness, Other] Presumably, this affects iOS GPS algorithm. **@see** [Apple docs](https://developer.apple.com/library/ios/documentation/CoreLocation/Reference/CLLocationManager_Class/CLLocationManager/CLLocationManager.html#//apple_ref/occ/instp/CLLocationManager/activityType) for more information |
| `pauseLocationUpdates`    | `Boolean`         | iOS          | Pauses location updates when app is paused (default: true)                                                                                                                                                                                                                                                                                         |
| `url`                     | `String`          | all          | Server url where to send HTTP POST with recorded locations **@see** [HTTP locations posting](#http-locations-posting)                                                                                                                                                                                                                              |
| `syncUrl`                 | `String`          | all          | Server url where to send fail to post locations **@see** [HTTP locations posting](#http-locations-posting)                                                                                                                                                                                                                                         |
| `syncThreshold`           | `Number`          | all          | Specifies how many previously failed locations will be sent to server at once (default: 100)                                                                                                                                                                                                                                                       |
| `httpHeaders`             | `Object`          | all          | Optional HTTP headers sent along in HTTP request                                                                                                                                                                                                                                                                                                   |
| `saveBatteryOnBackground` | `Boolean`         | iOS          | Switch to less accurate significant changes and region monitory when in background (default)                                                                                                                                                                                                                                                       |
| `maxLocations`            | `Number`          | all          | Limit maximum number of locations stored into db (default: 10000)                                                                                                                                                                                                                                                                                  |

Following options are specific to provider as defined by locationProvider option
### ANDROID_ACTIVITY_PROVIDER provider options

| Parameter             | Type      | Platform | Description                                                                                                                                                                                                                      |
|-----------------------|-----------|----------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| `interval`            | `Number`  | Android  | Rate in milliseconds at which your app prefers to receive location updates. **@see** [android docs](https://developers.google.com/android/reference/com/google/android/gms/location/LocationRequest.html#getInterval())          |
| `fastestInterval`     | `Number`  | Android  | Fastest rate in milliseconds at which your app can handle location updates. **@see** [android  docs](https://developers.google.com/android/reference/com/google/android/gms/location/LocationRequest.html#getFastestInterval()). |
| `activitiesInterval`  | `Number`  | Android  | Rate in milliseconds at which activity recognition occurs. Larger values will result in fewer activity detections while improving battery life.                                                                                  |
| `stopOnStillActivity` | `Boolean` | Android  | stop() is forced, when the STILL activity is detected (default is true)                                                                                                                                                          |

Success callback will be called with one argument - location object, which tries to mimic w3c [Coordinates interface](http://dev.w3.org/geo/api/spec-source.html#coordinates_interface).

| Callback parameter | Type      | Description                                                            |
|--------------------|-----------|------------------------------------------------------------------------|
| `locationId`       | `Number`  | ID of location as stored in DB (or null)                               |
| `provider`         | `String`  | gps, network, passive or fused                                         |
| `locationProvider` | `Number`  | Location provider                                                      |
| `debug`            | `Boolean` | true if location recorded as part of debug                             |
| `time`             | `Number`  | UTC time of this fix, in milliseconds since January 1, 1970.           |
| `latitude`         | `Number`  | latitude, in degrees.                                                  |
| `longitude`        | `Number`  | longitude, in degrees.                                                 |
| `accuracy`         | `Number`  | estimated accuracy of this location, in meters.                        |
| `speed`            | `Number`  | speed if it is available, in meters/second over ground.                |
| `altitude`         | `Number`  | altitude if available, in meters above the WGS 84 reference ellipsoid. |
| `bearing`          | `Number`  | bearing, in degrees.                                                   |


### start()
Platform: iOS, Android

Start background geolocation.

### stop()
Platform: iOS, Android

Stop background geolocation.

### isLocationEnabled(success, fail)

CHANGES.md
## Changelog

### [2.4.0] - 2017-1-06
### Fix
- iOS bugs #324, #294
### Add
- iOS LOCATION_ALWAYS_AND_WHEN_IN_USE_USAGE_DESCRIPTION in plugin.xml to support background location tracking in iOS-11

### [2.3.1] - 2017-10-31
### Fix
- iOS httpHeaders values are not sent with syncUrl on iOS PR #325

### [2.3.0] - 2017-10-31
### Added
- Android Make account name configurable PR #334 by unixmonkey

### [2.2.5] - 2016-11-13
### Fixed
- Android fixing issue #195 PR204

### [2.2.4] - 2016-09-24
### Fixed
- iOS extremely stupid config bug from 2.2.3

### [2.2.3] - 2016-09-23
### Fixed
- Android issue #173 - allow stop service and prevent crash on destroy

### [2.2.2] - 2016-09-22
### Added
- Android android.hardware.location permission

### Fixed
- iOS onStationary null location
- iOS fix potential issue sending outdated location
- iOS handle null config options

### [2.2.1] - 2016-09-15
### Added
- iOS suppress minor error messages on first app run

### [2.2.0] - 2016-09-14
### Added
- iOS option pauseLocationUpdates PR #156

### [2.2.0-alpha.8] - 2016-09-02
### Fixed
- iOS compilation errors

### [2.2.0-alpha.7] - 2016-09-01
#### Removed
- Android location filtering

### Changed
- Android db logging instead of file
- iOS location prop heading renamed to bearing

### [2.2.0-alpha.6] - 2016-08-10
### Fixed
- Android don't try sync when locations count is lower then threshold

### [2.2.0-alpha.5] - 2016-08-10
### Fixed
- Android issue #130 - sync complete notification stays visible
- Android don't try sync when locations count is zero

### [2.2.0-alpha.4] - 2016-08-10
### Fixed
- Android issue #137 - fix only for API LEVEL >= 17

### [2.2.0-alpha.3] - 2016-08-10
### Fixed
- Android issue #139 - Starting backgroundGeolocation just after configure failed

### [2.2.0-alpha.2] - 2016-08-10
### Fixed
- iOS issue #132 use Library as DB path

### [2.2.0-alpha.1] - 2016-08-01
### Added
- Android, iOS limit maximum number of locations in db (maxLocations)
- Android showAppSettings
- Android, iOS database logging (getLogEntries)
- Android, iOS autosync locations to server with configurable threshold
- Android, iOS method getValidLocations
- iOS watchLocationMode and stopWatchingLocationMode
- iOS configurable NSLocationAlwaysUsageDescription

### Changed
- Locations stored into db at all times
- iOS persist locations also when url option is not used
- iOS dropping support for iOS < 4

### Fixed
- Android fix crash on permission change
- Android permission error code: 2
- Android on start err callback instead configure err callback
- Android overall background service reliability
- iOS do not block js thread when posting locations

### [2.1.2] - 2016-06-23
### Fixed
- iOS database not created

### [2.1.1] - private release
### Fixed
- iOS switching mode

### [2.1.0] - private release
### Added
- iOS option saveBatteryOnBackground
- iOS time validation rule for location

### [2.0.0] - 2016-06-17
### Fixed
- iOS prevent unintentional start when in background
- Android Destroy Existing Provider Before Creating New One (#94)

### [2.0.0-rc.3] - 2016-06-13
#### Fixed
- iOS memory leak

### [2.0.0-rc.1] - 2016-06-13
#### Changed
- Android notificationIcon option split into small and large!!!
- Android stopOnTerminate defaults to true
- Android option locationService renamed to locationProvider
- Android providers renamed (see README.md)
- Android bugfixing
- SampleApp moved into separate repo
- deprecated backgroundGeoLocation
- iOS split cordova specific code to allow code sharing with react-native-background-geolocation
- desiredAccuracy map any number
- Android locationTimeout option renamed to interval
- iOS switchMode (formerly setPace)

#### Added
- Android startOnBoot option
- Android startForeground option
- iOS, Android http posting of locations (options url and httpHeaders)
- iOS showLocationSettings
- iOS showAppSettings
- iOS isLocationEnabled
- iOS getLocations
- iOS deleteLocation
- iOS deleteAllLocations
- iOS foreground mode

#### Removed
- WP8 platform
- Android deprecated window.plugins.backgroundGeoLocation

### [1.0.2] - 2016-06-09
#### Fixed
- iOS queued locations are send FIFO (before fix LIFO)

### [1.0.1] - 2016-06-03
#### Fixed
- iOS7 crash on start
- iOS attempt to fix #46 and #39

### [1.0.0] - 2016-06-01
#### Added
- Android ANDROID_FUSED_LOCATION stopOnStillActivity (enhancement #69)

### [0.9.6] - 2016-04-07
#### Fixed
- Android ANDROID_FUSED_LOCATION fixing crash on start
- Android ANDROID_FUSED_LOCATION unregisterReceiver on destroy

### [0.9.5] - 2016-04-05
#### Fixed
- Android ANDROID_FUSED_LOCATION startTracking when STILL after app has started

### [0.9.4] - 2016-01-31
#### Fixed
- Android 6.0 permissions issue #21

### [0.9.3] - 2016-01-29
#### Fixed
- iOS cordova 6 compilation error
- iOS fix for iOS 9

#### Changes
- iOS removing cordova-plugin-geolocation dependency
- iOS user prompt for using location services
- iOS error callback when location services are disabled
- iOS error callback when user denied location tracking
- iOS adding error callbacks to SampleApp

### [0.9.2] - 2016-01-29
#### Fixed
- iOS temporarily using cordova-plugin-geolocation-ios9-fix to fix issues with iOS9
- iOS fixing SampleApp indexedDB issues

### [0.9.1] - 2015-12-18
#### Fixed
- Android ANDROID_FUSED_LOCATION fix config setActivitiesInterval

### [0.9.0] - 2015-12-18
#### Changed
- Android ANDROID_FUSED_LOCATION using ActivityRecognition (saving battery)

### [0.8.3] - 2015-12-18
#### Fixed
- Android fixing crash on exit

### [0.8.2] - 2015-12-18
#### Fixed
- Android fixing #9 - immediate bg service crash

### [0.8.1] - 2015-12-15
#### Fixed
- Android fixing #9

### [0.8.0] - 2015-12-15 (Merry XMas Edition :-)
#### Fixed
- Android persist location when main activity was killed

#### Changed
- Android persisting position when debug is on

### [0.7.3] - 2015-11-06
#### Fixed
- Android issue #11

### [0.7.2] - 2015-10-21
#### Fixed
- iOS fixing plugin dependencies (build)

www/app.js
var map,
    previousLocation,
    locationMarkers = [],
    stationaryCircles = [],
    currentLocationMarker,
    locationAccuracyCircle,
    path,
    userStartIntent,
    isStarted = false,
    isLocationEnabled = false,
    configHasChanges = false;

var bgOptions = {
    stationaryRadius: 1,
    distanceFilter: 30,
    desiredAccuracy: 10,
    debug: true,
    notificationTitle: 'Background tracking',
    notificationText: 'enabled',
    notificationIconColor: '#FEDD1E',
    notificationIconLarge: 'mappointer_large',
    notificationIconSmall: 'mappointer_small',
    locationProvider: 0,//backgroundGeolocation.provider.ANDROID_DISTANCE_FILTER_PROVIDER,
    interval: 10,
    fastestInterval: 5,
    activitiesInterval: 10,
    stopOnTerminate: false,
    startOnBoot: true,
    startForeground: true,
    stopOnStillActivity: false,
    activityType: 'Other',
    //url: 'http://192.168.168.136/v1/geolocation',//brianserver
    url: 'http://192.168.1.9/v1/geolocation',//server1234
    // url: 'http://192.168.1.3/v1/geolocation',//brianmsi
    //syncUrl: 'http://192.168.3.185:3000/sync',
    syncThreshold: 10,
    commuter_id: 123789,
    trip_id: 'test1234',
    start_lat: 999998,
    start_lng: 999997,
end_lat: 51.5099,//london uk
end_lng: 0.1337,//london uk
    httpHeaders: {
        'X-FOO': 'bar'
    },
    pauseLocationUpdates: false,
    saveBatteryOnBackground: false,
    maxLocations: 100
};

var mapOptions = {
    center: { lat: 32.3318907, lng: -122.0318303 },
    zoom: 12,
    disableDefaultUI: false
};

try {
    Object.assign(bgOptions, JSON.parse(localStorage.getItem('bgOptions')));
}
catch (err) {
    console.log(err.message);
}


// Initialize app
var myApp = new Framework7({
    init: false,
    animateNavBackIcon: true,
    precompileTemplates: true,
    domCache: true,
    material: window.isAndroid,
    // fastclick: false
});

// We need to use custom DOM library, let's save it to $$ variable:
var $$ = Dom7;

var mainView = myApp.addView('.view-main');
var view1 = myApp.addView('#view-1');
var view2 = myApp.addView('#view-2');

myApp.onPageInit('map', function (page) {
    renderTabBar(isStarted);

    if (typeof google !== 'undefined') {
        map = new google.maps.Map(Dom7('#mapcanvas')[0], mapOptions);
    }

    if (typeof backgroundGeolocation === 'undefined') {
        myApp.alert('Plugin has not been initialized properly!', 'Error');
        return;
    }

    bgConfigure();
    backgroundGeolocation.onStationary(setStationary);
    backgroundGeolocation.watchLocationMode(
        function (enabled) {
            isLocationEnabled = enabled;
            if (enabled && userStartIntent) {
                startTracking();
            }
            else if (isStarted) {
                stopTracking();
                myApp.alert('Location tracking has been stopped');
            }
        },
        function (error) {
            myApp.alert(error, 'Error watching location mode');
        }
    );

    $$('#tabbar').on('click', '[data-action="tracking"]', function() {
        userStartIntent = !(isStarted & userStartIntent);
        toggleTracking(userStartIntent);
    });

});

myApp.onPageInit('settings', function (page) {
    var options = Object.assign({}, bgOptions);
    var locationProviders = [
        { name: 'ANDROID_DISTANCE_FILTER_PROVIDER', value: 0, selected: false, index: 0 },
        { name: 'ANDROID_ACTIVITY_PROVIDER', value: 1, selected: false, index: 1 },
    ];
    var selectedProvider = 0;

    if (options.locationProvider) {
        selectedProvider = Number(options.locationProvider);
        locationProviders[Number(options.locationProvider)].selected = true;
    }
    options.locationProvider = locationProviders[selectedProvider].name;
    options.locationProviders = locationProviders;

    $$('#settings').html(Template7.templates.settingsTemplate(options));

    $$('[data-action="back"]').click(function(ev) {
        var config = Array.prototype.reduce.call(
            $$('[data-page="settings"] [data-type="config"]'),
            function(values, el) {
                if (el.type === 'checkbox') {
                    values[el.name] = el.checked;
                }
                else {
                    if(!isNaN(parseInt(el.value))) {
                        values[el.name] = parseInt(el.value);
                    }
                    else {
                        values[el.name] = el.value;
                    }
                }
                return values;
            },
        {});
        bgConfigure(config);
        configHasChanges = false;
    });
});

$$('[data-page="settings"]').on('keyup keydown change', '[data-type="config"]', function(ev) {
    console.log('changed', this.name, this.checked, this.value);
    configHasChanges = true;
});

function toggleTracking(shouldStart) {
    if (shouldStart) {
        startTracking();
    }
    else {
        stopTracking();
    }
}

function bgConfigure(config) {
    Object.assign(bgOptions, config);
    localStorage.setItem('bgOptions', JSON.stringify(bgOptions));

    var options = Object.assign({}, bgOptions);
    if (options.interval) { options.interval *= 1000; }
    if (options.fastestInterval) { options.fastestInterval *= 1000; }
    if (options.activitiesInterval) { options.activitiesInterval *= 1000; }

    if (isStarted) {
        stopTracking();
        backgroundGeolocation.configure(
            setCurrentLocation,
            function (err) { console.log('Error occured', err); },
            options
        );
        startTracking();
    }
    else {
        backgroundGeolocation.configure(
            setCurrentLocation,
            function (err) { console.log('Error occured', err); },
            options
        );
    }
}

function startTracking() {
    if (isStarted) { return; }

    backgroundGeolocation.isLocationEnabled (
    function (enabled) {
        isLocationEnabled = enabled;
        if (enabled) {
            backgroundGeolocation.start(
                null,
                function (error) {
                    stopTracking();
                    if (error.code === 2) {
                        myApp.confirm('Would you like to open app settings?', 'Permission denied', function() {
                            backgroundGeolocation.showAppSettings();
                        });
                    }
                    else {
                        myApp.alert(error.message, 'Start failed');
                    }
                }
            );
            isStarted = true;
            renderTabBar(isStarted);
            }
            else {
                myApp.confirm('Would you like to open settings?', 'Location Services are disabled', function() {
                    backgroundGeolocation.showLocationSettings();
                });
            }
        },
        function (error) {
            myApp.alert(error, 'Error detecting status of location settings');
        }
    );
}

function stopTracking() {
    if (!isStarted) { return; }

    backgroundGeolocation.stop();
    isStarted = false;
    renderTabBar(isStarted);
}

function renderTabBar(isStarted) {
    $$('#tabbar').html(Template7.templates.tabbarTemplate({isStarted: isStarted}));
}

function setStationary (location) {
    console.log('[DEBUG] stationary recieved', location);
    var latlng = new google.maps.LatLng(Number(location.latitude), Number(location.longitude));
    var stationaryCircle = new google.maps.Circle({
        fillColor: 'pink',
        fillOpacity: 0.4,
        strokeOpacity: 0,
        map: map,
    });
    stationaryCircle.setCenter(latlng);
    stationaryCircle.setRadius(location.radius);
    stationaryCircles.push(stationaryCircle);
    backgroundGeoLocation.finish();
}

function setCurrentLocation (location) {
    console.log('[DEBUG] location recieved', location);
    if (!currentLocationMarker) {
        currentLocationMarker = new google.maps.Marker({
            map: map,
            icon: {
                path: google.maps.SymbolPath.CIRCLE,
                scale: 10,
                fillColor: '#ec0070', //'gold',
                fillOpacity: 1,
                strokeColor: '#22e2c5',
                strokeWeight: 3
            }
        });
        locationAccuracyCircle = new google.maps.Circle({
            fillColor: 'purple',
            fillOpacity: 0.4,
            strokeOpacity: 0,
            map: map
        });
    }
    if (!path) {
        path = new google.maps.Polyline({
            map: map,
            strokeColor: 'blue',
            fillOpacity: 0.4
        });
    }
    var latlng = new google.maps.LatLng(Number(location.latitude), Number(location.longitude));

    if (previousLocation) {
        // Drop a breadcrumb of where we've been.
        locationMarkers.push(new google.maps.Marker({
            icon: {
                path: google.maps.SymbolPath.CIRCLE,
                scale: 7,
                fillColor: 'green',
                fillOpacity: 1,
                strokeColor: 'white',
                strokeWeight: 3
            },
            map: map,
            position: new google.maps.LatLng(previousLocation.latitude, previousLocation.longitude)
        }));
    } else {
        if (map.getZoom() < 15) {
            map.setZoom(15);
        }
    }
    map.setCenter(latlng);

    // Update our current position marker and accuracy bubble.
    currentLocationMarker.setPosition(latlng);
    locationAccuracyCircle.setCenter(latlng);
    locationAccuracyCircle.setRadius(location.accuracy);

    // Add breadcrumb to current Polyline path.
    path.getPath().push(latlng);
    previousLocation = location;

    backgroundGeoLocation.finish();
}

function onDeviceReady() {
    backgroundGeolocation = window.backgroundGeolocation || window.backgroundGeoLocation || window.universalGeolocation;
    backgroundGeolocation.getLocations(function(locs) {
        var now = Date.now();
        var sameDayDiffInMillis = 24 * 3600 * 1000;
        locs.forEach(function (loc) {
            if ((now - loc.time) <= sameDayDiffInMillis) {
                setCurrentLocation(loc);
            }
        });
    });
    myApp.init();
}

document.addEventListener('deviceready', onDeviceReady, false);
